bftengine/src/bcstatetransfer/RangeValidationTree.cpp

```diff
@@ -127,116 +171,178 @@ void RangeValidationTree::RVTMetadata::staticAssert() noexcept {
 }
 
+// level 1 and up
 RVBNode::RVBNode(uint8_t level, uint64_t rvb_index)
-    : info_(level, rvb_index),
+    : info_(level, NodeInfo::minPossibleSiblingRvbIndex(rvb_index, level - 1)),
       current_value_(
           computeNodeInitialValue(info_, NodeVal::initialValueZeroData.data(), NodeVal::kDigestContextOutputSize),
-          NodeVal::kDigestContextOutputSize) {}
+          NodeVal::kDigestContextOutputSize) {
+  ConcordAssertGE(level, kDefaultRVTLeafLevel);
+  // logInfoVal("construct: "); // leave for debugging
+}
 
-RVBNode::RVBNode(uint64_t node_id, char* val_ptr, size_t size) : info_(node_id), current_value_(val_ptr, size) {}
+// serialized in memory nodes
+RVBNode::RVBNode(uint64_t node_id, char* val_ptr, size_t size) : info_(node_id), current_value_(val_ptr, size) {
+  ConcordAssertGE(info_.level(), kDefaultRVTLeafLevel);
+  // logInfoVal("construct: "); // leave for debugging
+}
 
-RVTNode::RVTNode(const shared_ptr<RVBNode>& node)
-    : RVBNode(kDefaultRVTLeafLevel, node->info_.rvb_index),
-      min_child_id{node->info_.rvb_index},
-      max_child_id{node->info_.rvb_index + RVT_K - 1},
+// Level 1 RVT Node ctor
+RVTNode::RVTNode(const shared_ptr<RVBNode>& child_node)
+    : RVBNode(kDefaultRVTLeafLevel, child_node->info_.rvb_index()),
+      parent_id_{},
+      insertion_counter_{},
       initial_value_(current_value_) {
-  n_child++;
+  auto min_sib_rvb_index =
+      NodeInfo::minPossibleSiblingRvbIndex(child_node->info_.rvb_index(), child_node->info_.level());
+  auto pow_child = pow_uint(RangeValidationTree::RVT_K, child_node->info_.level());
+  insertion_counter_ = (child_node->info_.rvb_index() - min_sib_rvb_index) / pow_child;
+
+  // TODO - add extra validatations here and inside child_ids child id and correct index
+  pushChildId(child_node->info_.id());
 }
 
-RVTNode::RVTNode(const RVTNodePtr& node)
-    : RVBNode(node->info_.level + 1, node->info_.rvb_index), initial_value_(current_value_) {
-  n_child++;
-  auto level = node->info_.level;
-  min_child_id = node->info_.id;
-  uint64_t rvb_index = node->info_.rvb_index + RangeValidationTree::pow_uint(RVT_K, level + 1) -
-                       RangeValidationTree::pow_uint(RVT_K, level);
-  max_child_id = NodeInfo(level, rvb_index).id;
-  ConcordAssert(n_child <= RVT_K);
-  // logInfoVal("construct: ");
+// level 2 and up nodes
+RVTNode::RVTNode(const RVTNodePtr& child_node)
+    : RVBNode(child_node->info_.level() + 1, child_node->info_.rvb_index()),
+      parent_id_{},
+      insertion_counter_{},
+      initial_value_(current_value_) {
+  // In some cases RVB node might be inserted in as a middle child (see more details in above ctor)
+  // The next calculation was verified with many examples.
+  // Here is one:
+  // RVT_K=12 and level 0 child is inserted on level 3 with an rvb_index 3000 (first insertion to tree)
+  // We know that current parent has the potential next child span:
+  // 13*12^2+1, 14*12^+1, ... [until 12th child] 23*12^2+1
```

Comment:
From where exactly range 13 to 23 came is not mentioned.
typo: 14*12^2+1

Reply:
3000 falls in that range (span) and there is a  way to calculate it.
1 line under :
`
 1729	1873	2017	2161	2305	2449	2593	2737	2881	3025	3169	3313
`